src/KDMqtt/mqtt.h

```diff
@@ -0,0 +1,250 @@
+
+  Contact KDAB at <info@kdab.com> for commercial licensing options.
+*/
+#pragma once
+
+#include <KDFoundation/file_descriptor_notifier.h>
+#include <KDFoundation/timer.h>
+#include <KDMqtt/kdmqtt_global.h>
+#include <KDMqtt/mosquitto_wrapper.h>
+#include <KDUtils/file.h>
+#include <KDUtils/url.h>
+#include <unordered_map>
+
+using namespace KDFoundation;
+using namespace KDUtils;
+
+namespace KDMqtt {
+
+constexpr int c_defaultPort = 1883;
+constexpr int c_defaultKeepAliveSeconds = 60;
+
+/*
+ * Class: IMqttLib
+ *
+ * This is an abstract class specifing a generic
+ * interface exposed to application / business logic
+ * to access MQTT library implementations.
+ */
+class KDMQTT_API IMqttLib
+{
+protected:
+    IMqttLib() = default;
+    virtual ~IMqttLib() { }
+
+public:
+    IMqttLib(const IMqttLib &) = delete;
+    IMqttLib &operator=(const IMqttLib &) = delete;
+
+    virtual int init() = 0;
+    virtual int cleanup() = 0;
+
+    [[nodiscard]] virtual bool isInitialized() const = 0;
+    virtual bool isValidTopicNameForSubscription(const std::string &topic) = 0;
+};
+
+/*
+ * Class: MqttLib
+ *
+ * This class exposes mosquitto library functions to
+ * the application / business logic.
+ */
+class KDMQTT_API MqttLib : public IMqttLib
+{
+    friend class MqttClient;
+    friend class MqttUnitTestHarness;
+
+private:
+    MqttLib();
+    ~MqttLib() = default;
+
+public:
+    static MqttLib &instance();
+
+    int init() override;
+    int cleanup() override;
+
+    [[nodiscard]] bool isInitialized() const override;
+    bool isValidTopicNameForSubscription(const std::string &topic) override;
+
+protected:
+    int version(int *major, int *minor, int *revision);
+
+    bool checkMosquittoResultAndDoDebugPrints(int result, std::string_view func = {});
+
+    std::string_view connackString(int connackCode);
+    std::string_view errorString(int errorCode);
+    std::string_view reasonString(int reasonCode);
+
+private:
+    MosquittoLib *m_mosquittoLib;
+    bool m_isInitialized;
+};
+
+/*
+ * Class: IMqttClient
+ *
+ * This is an abstract class specifing a generic
+ * interface exposed to application / business logic
+ * to access MQTT client implementations.
+ */
+class KDMQTT_API IMqttClient
+{
+public:
+    enum class ConnectionState {
+        CONNECTING,
+        CONNECTED,
+        DISCONNECTING,
+        DISCONNECTED
+    };
+
+    enum class SubscriptionState {
+        SUBSCRIBING,
+        SUBSCRIBED,
+        UNSUBSCRIBING,
+        UNSUBSCRIBED
+    };
+
+    KDBindings::Property<ConnectionState> connectionState{ ConnectionState::DISCONNECTED };
+    KDBindings::Property<SubscriptionState> subscriptionState{ SubscriptionState::UNSUBSCRIBED };
+
+    KDBindings::Property<std::vector<std::string>> subscriptions{};
+
+    KDBindings::Signal<int /*msgId*/> msgPublished;
+    KDBindings::Signal<const mosquitto_message * /*msg*/> msgReceived;
+
+    KDBindings::Signal<> error;
+
+    virtual int setTls(const File &cafile) = 0;
```

Comment:
We should also add the function that lets the user use the system certificate store for TLS.

I think that by default, we should use the system store by calling `mosquitto_int_option` with the `MOSQ_OPT_TLS_USE_OS_CERTS` option and let the user opt out of it somehow.

But also, remember to document that it only works on Linux because on Windows, OpenSSL used by mosquitto doesn't use the system store by default.